### ISSUE TITLE
eflags issue on x86_64

### ISSUE THREAD

**Issue by jbremer** [unicorn-engine/unicorn#496]
Hi all,

Excellent framework. Found the opportunity to take advantage of it in the past few days, overall worked pretty well. Ran into one pretty awful bug, though. Namely that the `eflags` register is not correctly updated in certain cases. This is a Unicorn-only bug (as in, not in upstream Qemu).

First of all I modified a couple of things regarding the Python bindings of which I'd be happy to see merged upstream - if you're willing to take these changes (please let me know).
- Some code cleanup to improve code style (https://github.com/jbremer/unicorn/commit/b7d817d107790c0473abc787aa49d6145c40a356)
- Simplification of register access (x86-only currently). This allows one to read the `eax` register through `mu.eax`, write the `r8` register through `mu.r8 = 1234`, etc (https://github.com/jbremer/unicorn/commit/ecb8fd3b6160dd30b5b9eded84375d2a82bd5a97)
- Return the amount of executed instructions if desired; this functionality is currently not yet very well designed as it only works when providing a maximum number of instructions to be executed (i.e., `mu.emu_start(start, end, count=-1 % 2**32)` - mostly because that functionality almost directly results in exposing the amount of executed instructions (at the cost of inserting a `UC_HOOK_CODE` in the background, of course), see also https://github.com/jbremer/unicorn/commit/fbb82f5fa18ca521b7dc67faff21a670e9207622
- Basic register import/export (Python binding only); this is just a basic wrapper for reading and writing all of the general purpose registers at once https://github.com/jbremer/unicorn/commit/dafc95a2df1818dc3b0e707436d2ea8986812757 https://github.com/jbremer/unicorn/commit/a65c6c06e49ae6b320c7bb7adb1293de57a5d99e

This brings us to the actual bug. It can be identified through the following script https://gist.github.com/jbremer/6c0aa8253304f7f5b385. Simply by commenting the `hook_add` one can _make this script work_. Obviously commenting that line out removes all calls to the `dumper` callback and also - in the case of counting executed instructions - would get rid of that underlying functionality.

So the expected outcome of the script is that in the third iteration, when the value at `[ebp-8]` is `0xc7`, the `cmp al, 0x19` instruction will set the `ZF` eflags bit (as `al` equals `0x19` in that case). It can be verified that when any `UC_HOOK_CODE` is enabled this does not work properly and instead ignores the conditional jump when `[ebp-8]` is set to `0xe0` - because the `sub` instruction set the `ZF` eflags bit.

Now interestingly enough I was able to identify this bug only hits when the `ZF` flag was already set. That is, when `ZF` is false, the `cmp` instruction sets it properly (this can be tested by uncommenting the `mu.eflags = 0xffffffff` line - don't forget to also test with `mu.eflags  = 0`). Now I didn't dig too much into the `qemu tcg` code, but as you may have already guessed, the bug is related to the `UC_HOOK_CODE` support in `qemu/target-i386/translate.c` and in fact reverting the commit from #448 fixed the issue for me. Why and how I'll leave up to you guys though.

Let me know if this is enough information :-) Thanks!

Regards,
Jurriaan

**Comment by lunixbochs** [comment]
I have mixed feelings on the Python register changes for a couple of reasons.

1. I like uc.regname, but I don't like the get/setattr indirection performance hit, and I don't like the multiple ways of doing the same thing. There was talk of a second Python binding with a different (simpler) interface, which I'd totally support for this.
2. x86-only, but this can be updated so it's not a huge deal.
3. I'm 100% against passing/returning a completely fixed list of registers, as it's impossible to add/move registers without breaking the API. A named tuple or just a class object is probably a better representation.

Another note - #8 talks about batch register reading in a more efficient way (if the C API supports it).

---

As far as the bug: if TCG sometimes keeps EFLAGS in a temporary register and doesn't always update the CPU struct, we'd need a "save eflags" op generated before the hook, otherwise the "restore eflags" code could load an old value.

**Comment by aquynh** [comment]
this is interesting! while we discuss all these ideas, please send a pull request on improving Python code style (https://github.com/jbremer/unicorn/commit/b7d817d107790c0473abc787aa49d6145c40a356), as that can be independently merged first.

thanks.

**Comment by lunixbochs** [comment]
I haven't noticed any problems on x86-64, but EFLAGS is definitely broken by HOOK_CODE on x86-32.

**Comment by jbremer** [comment]
@lunixbochs Agreed. Personally I thought the register wrappers would definitely be interesting for upstream Unicorn. The performance hit of a single function wrapper compared to calling a C library using the `ctypes`/`ffi` style of doing things is farily negligible, afaik.

The other two features were mostly nice-to-haves that I needed myself; I definitely agree that the register import/export is not production ready as-is, but both import/export and the instruction counting are features that could be useful to have - whether this approach is taken or whether a more *clean* implementation is taken, is up to you guys. I guess :-)

If there is already a new Python API upcoming that's also totally fine of course, but perhaps these commits could propose some ideas / functions for that new API layer.

Thanks again!

**Comment by lunixbochs** [comment]
I measured it last time: https://github.com/unicorn-engine/unicorn/issues/201#issuecomment-177604294

**Comment by jbremer** [comment]
This is completely different from what I proposed in my commit, for a reason. Naturally I foresaw this performance hit ;-)
What that gist does is for *every* register read/write it creates the `UC_X86_REG_XYZ` string, looks it up in a global variable, and then actually read/writes the register based on its integer value.
As you can see in my proposed commit https://github.com/jbremer/unicorn/commit/ecb8fd3b6160dd30b5b9eded84375d2a82bd5a97 it will do this step of looking up the correct indices *once* and then creates a `property()` wrapper based on that integer value - this is also the reason I wrapped the property generation in a separate function; so to make Python's scoping work properly.

If you have the time I'd be interested to see benchmarks on my commit as I assure you that I explicitly took care of this performance hit :-)

**Comment by lunixbochs** [comment]
('magic:', 2.9814491271972656)
    ('native:', 2.72104811668396)

But you still have the problems of getting the registers onto each Uc efficiently and the fact you're creating two ways to do something (which is why I still think this should be part of a separate API that doesn't mirror the C style)

**Comment by aquynh** [comment]
on the new API you are proposing to get the instruction counter, i rather extending uc_emu_start() to return that than having a separate API. this breaks the API compatibility, but we already broke that anyway.

**Comment by lunixbochs** [comment]
is https://github.com/unicorn-engine/unicorn/blob/master/qemu/target-i386/translate.c#L4781 not always syncing eflags a problem here?

**Comment by zachriggle** [comment]
I manually moved that line and the issue still persists as reported in #501.

```diff
diff --git a/qemu/target-i386/translate.c b/qemu/target-i386/translate.c
index 76bab2d..3e5170b 100644
--- a/qemu/target-i386/translate.c
+++ b/qemu/target-i386/translate.c
@@ -4778,8 +4778,8 @@ static target_ulong disas_insn(CPUX86State *env, DisasContext *s,
 
     // Unicorn: trace this instruction on request
     if (HOOK_EXISTS_BOUNDED(env->uc, UC_HOOK_CODE, pc_start)) {
+        sync_eflags(s, tcg_ctx);
         if (s->last_cc_op != s->cc_op) {
-            sync_eflags(s, tcg_ctx);
             s->last_cc_op = s->cc_op;
             changed_cc_op = true;
         }
```

**Comment by lunixbochs** [comment]
I move to disable the call to `restore_eflags` in HOOK_CODE until this is fixed, as it currently completely breaks execution (you'll just blow through jumps at random). PR #505

**Comment by zachriggle** [comment]
I agree with the removal.  It fixes my bug #499, and makes both #501 and #511 start passing.
